### PR TITLE
Add `--github-username` option and config to `bundle gem`

### PR DESCRIPTION
**Pull request by okuramasafumi**
# Description:

This option allows to fill in `[USERNAME]` in README with explicit value from user input.

Before this change, the value came only from `git config github.user`.
This can be a good way for some users, but there are a few problems.
Firstly, it's not documented that we can fill in `[USERNAME]` this way.
Secondly, the config value is set by some tools implicitly, and is not documented in `git-config` manpage.

Here are some changes in this Pull Request to address these issues.

* Add `--github-username` option to `bundle gem` command
* Save config value if a user gives `--github-username`
* Address the fact that we first look to `git config github.user` for GitHub username in command option description

Note:
This Pull Request actually doesn't change its current behavior.
If a user has `github.user` in their git config, it'll be used.
It's described in command description, but maybe it's not enough.

## What was the end-user or developer problem that led to this PR?

I sometimes create a new gem and there are always `[USERNAME]` in its generated README.
I didn't know that I can fill in by setting `github.user` in git config, and after knowing it I don't like that way since that value in config is not documented anywhere. I wanted more explicit way of doing so.

## What is your fix for the problem, implemented in this PR?

I added `--github-username` option to `bungle gem` command.

______________

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
> Firstly, it's not documented that we can fill in [USERNAME] this way.

I would say that the solution to this would be to document it?

> Secondly, the config value is set by some tools implicitly, and is not documented in git-config manpage.

Which tools set this? Maybe using a custom git config is not the best option here, but I don't like a CLI flag either since users will always want the same value for it. So configuration is better for this than a CLI flag that you have to specifiy every time.

I listen to ideas here, but my initial take on this is to document this and leave things as they are.

**Comment by okuramasafumi** [comment]
> I would say that the solution to this would be to document it?

Yes it should be a quick solution. I personally prefer putting settings about rubygems/bundler in its own config file over putting them in git config.

> Which tools set this?

I heard `hub` command sets this.

> since users will always want the same value for it. So configuration is better for this than a CLI flag that you have to specifiy every time.

I agree, so this PR adds the command line option but doesn't force users to fill in the value, and lets users store the value in bundler config after they fill it in for the first time so that they don't have to specify everytime.

**Comment by deivid-rodriguez** [comment]
@okuramasafumi I believe you have a point, and also you're right that this is not only a CLI flag but also a permanent configuration. So I'm ok with this in general.

My remaining doubt is whether it makes sense to "dirty" the CLI that will be only used once, and whether it's better to provide this as a configuration setting only. What do you think?

**Comment by deivid-rodriguez** [comment]
Something else I'm noticing, although not directly related to this PR, is that our gem generator assumes the code will be hosted on GitHub, so I believe we could provide a more useful default template by prefilling gemspec fields such as `homepage`, `homepage_uri`, `source_code_uri` and `changelog_uri` with "github values".

**Comment by hsbt** [comment]
> My remaining doubt is whether it makes sense to "dirty" the CLI that will be only used once, and whether it's better to provide this as a configuration setting only. What do you think?

I agree with this. It's better to add the alternative options to bundle config.

**Comment by okuramasafumi** [comment]
> My remaining doubt is whether it makes sense to "dirty" the CLI that will be only used once, and whether it's better to provide this as a configuration setting only. What do you think?

I see that although we might want to change our test framework/CI tool sometimes, we might never change our GitHub username. Adding a new flag for this option can be "dirty".

I'll make my PR like:

* Remove CLI flag
* Fetch config value from git config first not to break backward compatibility
* When config value in git config is missing, try to fetch the value from bundle config
* When they are all missing, default to `[USERNAME]`

**Comment by deivid-rodriguez** [comment]
Yeah, but I'm still in doubt because looking at where this is used (code hosting URLs essentially), you might want to create a gem under your employer's github organization, for example. So that would be a use case.

**Comment by okuramasafumi** [comment]
Ah that would be a use case. I think we can leave the CLI unchanged and if users want a CLI flag we can add it later.

**Comment by okuramasafumi** [comment]
Some specs fail in CI but not in my local env (I tried reusing seed but the result was the same).
It says `bundle!` command is not found, both in Ubuntu and Windows platform.
They happen in specs I added. However, there are the same `bundle!` method used in similar specs and they are all green.
I inspected and the only difference is that they have `readline` flag in its `describe`.
I don't know how to fix it. What can I do?

**Comment by deivid-rodriguez** [comment]
You need to rebase your PR on top of the latest master and replace `bundle!` with `bundle`, since we have removed the former.

**Comment by okuramasafumi** [comment]
@deivid-rodriguez Thanks, now all specs are green!
Should I squash three commits into one?

**Comment by deivid-rodriguez** [comment]
Don't squash commits yet, since I'm still thinking whether it's better to introduce `--github-username` as a flag directly. I plan to create a gem under the `rubygems` organization soon, and I think I would use the flag :sweat_smile:. Also, I believe this would be the first `gem.<configuration>` that doesn't have a corresponding CLI flag, so adding the floag might save users from some surprises. @hsbt Any thoughts?

File: bundler/lib/bundler/cli.rb [hunk 1]
```diff
@@ -577,7 +577,6 @@ def viz
                          :desc => "Generate a test directory for your library, either rspec, minitest or test-unit. Set a default with `bundle config set gem.test (rspec|minitest|test-unit)`."
     method_option :ci, :type => :string, :lazy_default => Bundler.settings["gem.ci"] || "",
                        :desc => "Generate CI configuration, either GitHub Actions, Travis CI, GitLab CI or CircleCI. Set a default with `bundle config set gem.ci (github|travis|gitlab|circle)`"
-
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli.rb:580]: Can you restore this blank line?

File: bundler/man/bundle-check.1.txt [hunk 2]
```diff
@@ -9,8 +9,8 @@ SYNOPSIS
        bundle check [--dry-run] [--gemfile=FILE] [--path=PATH]
 
 DESCRIPTION
-       check  searches the local machine for each of the gems requested in the
-       Gemfile. If all gems are found, Bundler prints a  success  message  and
+       check searches the local machine for each of the gems requested in  the
+       Gemfile.  If  all  gems are found, Bundler prints a success message and
```
Comment (deivid-rodriguez) [inline bundler/man/bundle-check.1.txt:13]: I hate when man pages updates do this... I should fix that.

File: bundler/man/bundle-config.ronn [hunk 3]
```diff
@@ -193,6 +193,9 @@ learn more about their operation in [bundle install(1)](bundle-install.1.html).
    Disallow changes to the `Gemfile`. When the `Gemfile` is changed and the
    lockfile has not been updated, running Bundler commands will be blocked.
    Defaults to `true` when `--deployment` is used.
+* `gem.github_username` (`BUNDLE_GEM__GITHUB_USERNAME`):
+   Sets a username for GitHub. This will be used in `README` file when you
```
Comment (deivid-rodriguez) [inline bundler/man/bundle-config.ronn:197]: I would say 
```suggestion
   Sets a GitHub username or organization to be used in `README` file when you
```

**Comment by deivid-rodriguez** [comment]
Any feedback on my latest comment? Should we go with the original `--github-username` flag approach?

**Comment by okuramasafumi** [comment]
@deivid-rodriguez Sorry for the late reply, I noticed that there are so many files to be rebased so I'll do that later.

For the spec of the flag, I'm not sure how many people add the flag just once.
If people need to add this flag many times to switch from personal to organization and vice versa, it doesn't make sense anymore.

**Comment by deivid-rodriguez** [comment]
> @deivid-rodriguez Sorry for the late reply, I noticed that there are so many files to be rebased so I'll do that later.

I can also do that for you, no problem.

> For the spec of the flag, I'm not sure how many people add the flag just once.
> If people need to add this flag many times to switch from personal to organization and vice versa, it doesn't make sense anymore.

Even if people need to add the flag many times, it's still useful, no? I mean, if you create 80% of gems under your organization's account, and 20% under your personal account, it still sounds like useful to have both a configuration and a CLI flag, so that you can configure `bundle gem` to normally use your organization, but also be able to pass `--github-username` when you want to create a gem under your personal account.

**Comment by okuramasafumi** [comment]
> if you create 80% of gems under your organization's account, and 20% under your personal account, it still sounds like useful to have both a configuration and a CLI flag, so that you can configure `bundle gem` to normally use your organization, but also be able to pass `--github-username` when you want to create a gem under your personal account.

I consider this situation and suspect that whether we should overwrite configuration value when users give the flag second time.
Maybe we should save the value from the flag to the config value only when the config value is empty. In other words, when users already have the value in config, the flag changes it temporarily, not parmanently.

**Comment by deivid-rodriguez** [comment]
Mmmmm, is that no how it currently works? If it's not, I think that's definitely how it should work!

**Comment by deivid-rodriguez** [comment]
I rebased the PR. If we decide to keep the flag, it'd be a matter of reverting the second commit.

**Comment by okuramasafumi** [comment]
Thank you!
One if the checks failed due to something related to RuboCop, but I could not resolve it :(

**Comment by deivid-rodriguez** [comment]
Sorry about that, it should be fixed by https://github.com/rubygems/rubygems/pull/3946. We can rebase again after that's merged.

**Comment by okuramasafumi** [comment]
Cool, and I'll add one more commit to make the flag to change the config temporarily if there's already the config value.

**Comment by deivid-rodriguez** [comment]
@okuramasafumi I double checked and this is the current behavior, so I don't think there's anything to change.

```
$ bundle config
Settings are listed in order of priority. The top value will be used.

$ bundle gem foo
Creating gem 'foo'...
Do you want to generate tests with your gem?
Future `bundle gem` calls will use your choice. This setting can be changed anytime with `bundle config gem.test`.
Enter a test framework. rspec/minitest/test-unit/(none): rspec
Do you want to set up continuous integration for your gem? Supported services:
* CircleCI:       https://circleci.com/
* GitHub Actions: https://github.com/features/actions
* GitLab CI:      https://docs.gitlab.com/ee/ci/
* Travis CI:      https://travis-ci.org/

Future `bundle gem` calls will use your choice. This setting can be changed anytime with `bundle config gem.ci`.
Enter a CI service. github/travis/gitlab/circle/(none): github
Do you want to license your code permissively under the MIT license?
This means that any other developer or company will be legally allowed to use your code for free as long as they admit you created it. You can read more about the MIT license at https://choosealicense.com/licenses/mit. y/(n): 
Do you want to include a code of conduct in gems you generate?
Codes of conduct can increase contributions to your project by contributors who prefer collaborative, safe spaces. You can read more about the code of conduct at contributor-covenant.org. Having a code of conduct means agreeing to the responsibility of enforcing it, so be sure that you are prepared to do that. Be sure that your email address is specified as a contact in the generated code of conduct so that people know who to contact in case of a violation. For suggestions about how to enforce codes of conduct, see https://bit.ly/coc-enforcement. y/(n): 
Do you want to add rubocop as a dependency for gems you generate?
RuboCop is a static code analyzer that has out-of-the-box rules for many of the guidelines in the community style guide. For more information, see the RuboCop docs (https://docs.rubocop.org/en/stable/) and the Ruby Style Guides (https://github.com/rubocop-hq/ruby-style-guide). y/(n): 
      create  foo/Gemfile
      create  foo/lib/foo.rb
      create  foo/lib/foo/version.rb
      create  foo/foo.gemspec
      create  foo/Rakefile
      create  foo/README.md
      create  foo/bin/console
      create  foo/bin/setup
      create  foo/.gitignore
      create  foo/.rspec
      create  foo/spec/spec_helper.rb
      create  foo/spec/foo_spec.rb
      create  foo/.github/workflows/main.yml
Initializing git repo in /home/deivid/Code/playground/foo
Gem 'foo' was successfully created. For more information on making a RubyGem visit https://bundler.io/guides/creating_gem.html

$ bundle config
Settings are listed in order of priority. The top value will be used.
gem.test
Set for the current user (/home/deivid/.bundle/config): "rspec"

gem.ci
Set for the current user (/home/deivid/.bundle/config): "github"

gem.mit
Set for the current user (/home/deivid/.bundle/config): false

gem.coc
Set for the current user (/home/deivid/.bundle/config): false

gem.rubocop
Set for the current user (/home/deivid/.bundle/config): false

$ bundle gem foo --test=minitest
Creating gem 'foo'...
    conflict  foo/Gemfile
Overwrite /home/deivid/Code/playground/foo/Gemfile? (enter "h" for help) [Ynaqdhm] deivid@chaba ~/Code/playground $ bundle gem foo2 --test=minitest
Creating gem 'foo2'...
      create  foo2/Gemfile
      create  foo2/lib/foo2.rb
      create  foo2/lib/foo2/version.rb
      create  foo2/foo2.gemspec
      create  foo2/Rakefile
      create  foo2/README.md
      create  foo2/bin/console
      create  foo2/bin/setup
      create  foo2/.gitignore
      create  foo2/test/test_helper.rb
      create  foo2/test/test_foo2.rb
      create  foo2/.github/workflows/main.yml
Initializing git repo in /home/deivid/Code/playground/foo2
Gem 'foo2' was successfully created. For more information on making a RubyGem visit https://bundler.io/guides/creating_gem.html

$ bundle config
Settings are listed in order of priority. The top value will be used.
gem.test
Set for the current user (/home/deivid/.bundle/config): "rspec"

gem.ci
Set for the current user (/home/deivid/.bundle/config): "github"

gem.mit
Set for the current user (/home/deivid/.bundle/config): false

gem.coc
Set for the current user (/home/deivid/.bundle/config): false

gem.rubocop
Set for the current user (/home/deivid/.bundle/config): false
```

**Comment by okuramasafumi** [comment]
Now I see. Then, should I revert the second commit?

**Comment by deivid-rodriguez** [comment]
Yeah, I think it's a valid use case after all, and it keeps things more consistent. Sorry for the extra work :pray:.

**Comment by deivid-rodriguez** [comment]
Just merged the PR I mentioned, so rebasing again should get this PR green.

**Comment by deivid-rodriguez** [comment]
There's still some issues in the specs, since we removed the `bundle!` helper. You should use `bundle` now.

**Comment by deivid-rodriguez** [comment]
Also, I'm reading the specs and I wonder. Shouldn't the new option/configuration take precedence over the old `github.user` git configuration?

**Comment by deivid-rodriguez** [comment]
@okuramasafumi Are you still interested in this PR?

**Comment by deivid-rodriguez** [comment]
Ok, I addressed all concerns myself, and this one should be ready now!

**Comment by okuramasafumi** [comment]
@deivid-rodriguez Thank you for your hard work!

**Pull request merged**